[Si Kai] iP

sikai00 opened the pull request:
## Dodie 🌻 
> [6/10](https://genius.com/Dodie-6-10-lyrics) pretty good bot

Dodie remembers your todo list so you don't have to. It's,

- text-based
- easy to learn
- ~fast~ SUPER FAST to learn and use

All you need to do,

1. Download it from [here](https://github.com/sikai00/ip/releases/tag/A-Jar)
2. Create a folder and place the downloaded JAR file in it (name the folder Dodie!)
3. Double-click on it **or, for advanced users, perform ```java -jar duke.jar``` in your terminal in the Dodie folder**
4. Let it manage your tasks for you 😌 

And it's **FREE**[^1]!
[^1]: No batteries included.

Features:
- [x] **3 different types of tasks**: Todo, Events and Deadlines
- [x] Find your task by keywords
- [x] add, delete, mark, unmark
- [x] List out all your tasks
- [x] GUI
- [ ] List view, sorted by task type
- [ ] Shortcuts to enter dates

Todo:
- [ ] Re-theme to Dodie 🌻 

Bugs:
- [ ] Overflow of text especially when using ```list``` is clipped by ellipsis

### If you're a Java programmer too...
You can change the directory of the storage by going into [here](src/main/java/duke/utils/Storage.java) and updating some arguments as previewed in this code block:
```java
 File dataDirectory = new File("data"); // directory
        if (!dataDirectory.exists()) {
            dataDirectory.mkdir();
        }

        try {
            FileWriter fw = new FileWriter("data/duke.txt", false); // actual file path
```

Commit:
Add Gradle support

sikai00 committed:
Duke.java: Add pretty print method to format all messages

sikai00 committed:
Duke.java: Complete Level 1. Greet, Echo, Exit

sikai00 committed:
Duke.java: Complete Level 2: Add, List

Refactored Duke into its own object

Duke now has instance level methods and attributes. This is primarily
done to allow better future code expansion.

Change usrCommand variable to usrInput

usrCommand is misleading as tasks are not commands. The use case of the
command line have expanded to include addition of new tasks without
commands but by simply entering the task into the command line.

sikai00 committed:
Change representation of tasks into its own class

This is a more natural approach to represent tasks. Task can also have
more information beyond its title for future code extension. A TaskList
class is also added to manage and store these tasks. These changes allow
for abstraction barriers to be erected.

sikai00 committed:
Complete Level 3. Mark as Done

sikai00 committed:
Add support for different types of task

sikai00 committed:
Task.java: Change to an abstract class

We are now dividing task into 3 different types. As such we change Task
into an abstract class since these different types are all Task. This is
to allow cleaner code and also to maintain our use of TaskList.

sikai00 committed:
TaskList.java: Add size method

sikai00 committed:
Complete Level 4. ToDos, Events, Deadlines

addTask now takes in Task instead of String.

This is to pass the logic of deciding what kind of a task to the main
logic loop. This is due to the different fields a task may have. Making
addTask handle such decisions can lead to many parameters needed and
hard to maintain.

Messages are now divided into start, content and end.

We now require intro messages as well as end messages for any command
input. Splitting message up will allow us to have greater readability.

sikai00 committed:
Duke.java: Fix and tidy up messages to match change in style

sikai00 committed:
Duke.java: Refactor main loop if-else into switch

Using if-else statements for main loop had bad readability. This is due to the denseness of the statements. Using switch statement will allow for better readability since the conditions are clear. One trade off is that the "bye" condition cannot exist within the switch statement due to the location of break. However, readability is more important and the "bye" condition is close enough to the switch statements.

sikai00 committed:
Complete A-TextUiTesting

After refactoring of code in the previous commit, the automated Text UI
testing was used to check for regressions.

sikai00 committed:
Change wording from title to description to match guide

sikai00 committed:
Complete Level 5. Handle Errors

Other changes:
Changed from 2 whitespaces for tabs to 4 as per guidelines.

sikai00 committed:
Complete Level 6. Delete

sikai00 committed:
Duke.java: Use getMessage() to print exception instead of toString()

sikai00 committed:
Duke.java: Make some methods throw exceptions

This is done as it is more natural for methods to check these inputs
rather than for the main application logic to check.

sikai00 committed:
Duke.java: Add message for no tasks remaining for  viewAllTask

sikai00 committed:
Update A-TextUiTesting to match new message

sikai00 committed:
Duke.java: Move greetings/good bye into their own method

sikai00 committed:
Complete A-Enums

Command.java: getIfPresent
This is implemented in order to handle any inputs that may be invalid.
This is because by default, Enum.valueOf throws an
IllegalArgumentException upon an invalid argument.

As it is likely that the input can be invalid and this is more of an
checked exception, let's use getIfPresent to handle wrong user inputs
and return Command.INVALID to signify wrong user inputs.

sikai00 committed:
Duke.java: Move runCommand into its own method

The entry point of Duke.java is overcrowded as command handling code are
all written in it. By moving runCommand into its own method, we are able
to better compartmentalize our code. We can move the main application
loop into its own method as well. This allows our entry point to simply
be the driver.

sikai00 committed:
Duke.java Format to fit character limit per line.

sikai00 committed:
EXPECTED.TXT: Add test for delete

sikai00 committed:
Duke.java: Handle NumberFormatException in parseInt

sikai00 committed:
Add test for invalid argument for DELETE

sikai00 committed:
Duke.java: Fix switch block to follow style

sikai00 committed:
Add methods to Task to retrieve task type and time

There is no way to get the task type easily and no way to get the time
of the event (whether it is deadline or event). These methods are
necessary as they are useful for parsing, especially in persistent
storage.

sikai00 committed:
Add Storage to handle persistent storage feature

sikai00 committed:
Complete Level 7. Save

Constructor no longer initializes a TaskList. It is more natural for
startDuke() to initialize a TaskList as a TaskList is only needed when
Duke is started. Furthermore, this allows us to add the logic to read
from storage in a cleaner fashion (inside startDuke).

Other changes:
addTask, deleteTask, markTask and unmarkTask will now make changes to
the persistent storage file using Storage's static methods.

sikai00 committed:
Complete Level 8. Dates and Times

Change date from string to LocalDateTime for all tasks with date
This allows us to read and parse dates easier, as well as print dates
using inbuilt formatters.

sikai00 committed:
Merge branch 'branch-Level-7'

sikai00 committed:
Duke.java: Normalise all messages to not have space after colon

sikai00 committed:
Update automated text-ui-text

sikai00 committed:
Merge branch 'branch-Level-8'

Change return type of getTime() to Optional<LocalDateTime>
Todo should not have a LocalDateTime. However, all task should have a
getTime method for cleaner code, especially when supporting persistent
storage. We want Todo to have a null LocalDateTime, hence Optional is
chosen to represent both null LocalDateTime and non-null.

sikai00 committed:
Complete A-MoreOOP

sikai00 committed:
Complete A-Packages

sikai00 committed:
Merge branch 'add-gradle-support'

sikai00 committed:
Complete A-Gradle. Use Gradle

sikai00 committed:
Complete A-JUnit. Add JUnit Tests

sikai00 committed:
Duke.java: Fix run()

sikai00 committed:
Complete A-JavaDoc. JavaDoc

sikai00 committed:
Complete A-CodingStandard. Follow the Coding Standard

sikai00 committed:
Complete Level 9. Find

sikai00 committed:
Merge branch 'branch-A-JavaDoc'

sikai00 committed:
Parser.java: Fix checkstyle error

sikai00 committed:
Merge branch 'branch-A-CodingStandard'

sikai00 committed:
Merge branch 'branch-Level-9'

sikai00 committed:
Add line break for JavaDoc between description and param

sikai00 committed:
Refactor Storage.java into instance-based

sikai00 committed:
Style code to comply with coding standard

sohenze commented on src/main/java/duke/utils/Storage.java:
```diff
@@ -0,0 +1,128 @@
+                String taskDescription = taskStrTokens[1];
+                boolean isTaskDone = taskStrTokens[2].equals("0") ? false : true;
+
+                switch (taskType) {
+                case "Todo":
+                    Todo currTodo = new Todo(taskDescription, isTaskDone);
+                    taskList.addTask(currTodo);
+                    break;
+                case "Deadline":
+                    LocalDateTime by = LocalDateTime.parse(taskStrTokens[3]);
+                    Deadline currDeadline = new Deadline(taskDescription, isTaskDone, by);
+                    taskList.addTask(currDeadline);
+                    break;
+                case "Event":
+                    LocalDateTime at = LocalDateTime.parse(taskStrTokens[3]);
+                    Event currEvent = new Event(taskDescription, isTaskDone, at);
+                    taskList.addTask(currEvent);
+                    break;
+                default:
+                    // Something that cannot be recognized
+                    break;
```
I like how you noticed that we should not use indentations for the case clauses

sohenze commented on src/main/java/duke/commands/AddCommand.java:
```diff
@@ -0,0 +1,41 @@
+ */
+public class AddCommand extends Command {
+    /** Command identifier used by Parser **/
+    public static final String COMMAND_WORD = "add";
```
I like how you use screaming snake case for a constant

sohenze commented on src/test/java/duke/utils/StorageTest.java:
```diff
@@ -0,0 +1,61 @@
+    private static final String FILE_NAME = "testStorage";
+    private static final String FILE_DIRECTORY = "storage/testStorage.txt";
+
+    // Credits: https://www.baeldung.com/java-delete-directory
```
I like how you credited the original source

sohenze commented on src/main/java/duke/utils/Parser.java:
```diff
@@ -0,0 +1,222 @@
+        return new FindCommand(tokens[1]);
+    }
+
+    /**
+     * Creates and returns LocalDate objects by parsing user's input for date.
+     * User input date should follow the format (yyyy-mm-dd HH:MM), in 24-hour time format.
+     * The date can have a delimiter of a character from "-/|" (characters within the double quotes).
+     * The time can have a delimiter of a character from ":.-|" (characters within the double quotes).
+     * The date and time must be separated by a single whitespace.
+     *
+     * @param dateText User's input of date and time following the format specified.
+     * @return LocalDateTime
+     */
```
I like how you have detailed header comments for all your public classes and methods

sikai00 committed:
Complete basic integration of GUI with JavaFX

Bugs:
No exit message after user inputs 'bye'
CLI-related artifacts still remains

sikai00 committed:
Complete Level 10. GUI

sikai00 committed:
Add credits for JavaFX Components

sikai00 committed:
Remove unused Ui files and related methods to them

sikai00 committed:
CommandResult.java: Add JavaDoc

sikai00 committed:
DialogBox.fxml: Add padding to label

sikai00 committed:
Main.java: Set application to have title

sikai00 reviewed (commented):
--accidental code review, removed--

sikai00 committed:
Merge branch 'branch-Level-10'

leewaichong commented on src/main/java/duke/tasks/Task.java:
```diff
@@ -0,0 +1,71 @@
+/**
+ * Represents a task.
+ */
+public abstract class Task {
```
I like how you made Task an abstract class. Very neat and clean trick.

leewaichong commented on src/main/java/duke/utils/Parser.java:
```diff
@@ -0,0 +1,222 @@
+        String[] tokens = userInput.split(" ", 2);
+        String command = tokens[0];
+
+        switch (command) {
```
I like how you use switch case instead of if else. It made the code very neat, succinct and very easy to follow.

leewaichong reviewed (commented):
Overall, I like how clean your code is and how well encapsulated everything is. There was not much repetition of code and everything that can be extracted out was made as methods. Another very neat trick was separating the task files into different directories respectively, making it very clear for others to follow and trace through. Good Job!

sikai00 committed:
Write test for findTask

sikai00 committed:
Use String.contains instead of regex

The findTask method uses regex to match for keywords. This is
unnecessary when there is a better and simpler method to perform the
same action.

Let's practice KISS and use String.contains to find keywords instead.

sikai00 committed:
Refactor code

sikai00 committed:
Replace use of DukeException with CommandResult

DukeException serves no real functional purpose with the creation of
CommandResult. It only messes up the code and makes it harder to
navigate and untidy. CommandResult can already encapsulate messages,
including both success and failure messages, to be sent to the users.
Since exceptions are used to handle user fault, it is okay for the user
to see the exception (which actually complies with previous behaviour).

Let's remove the use of DukeException and wrap error messages that the
user is supposed to see anyway within the CommandResult.

sikai00 committed:
Fix minor extra newline

sikai00 committed:
Complete A-Assertions. Use Assertions

sikai00 committed:
Complete A-Streams. Use Streams

sikai00 committed:
Merge pull request #1 from sikai00/branch-A-Assertions

Complete A-Assertions. Use Assertions

sikai00 committed:
Merge branch 'master' into branch-A-CodeQuality

sikai00 committed:
Merge pull request #2 from sikai00/branch-A-CodeQuality

Complete A-CodeQuality. Improve Code Quality

sikai00 committed:
Merge branch 'master' into branch-A-Streams

sikai00 committed:
Merge pull request #3 from sikai00/branch-A-Streams

Complete A-Streams. Use Streams

sikai00 committed:
Complete A-CI. Set up CI

sikai00 committed:
Add description to .gitignore

sikai00 committed:
Fix naming issue for GitHub actions

sikai00 committed:
Add view by today, tomorrow or week command

sikai00 committed:
Parser.java: Fix assertation error bug

sikai00 committed:
Merge branch 'branch-B-ViewSchedules'

sikai00 committed:
Move exit logic to within ExitCommand

The exit logic is weirdly wired through MainWindow which should not be
responsible for the effects of a Command execution.

Let's move the exit logic using TimerTask into ExitCommand to be
executed. This follows all other type of Commands' behaviours.

A minor note: As running Timer is asynchronous, we are able to achieve
this effect without much issues.

sikai00 committed:
Fix and abide by boolean naming convention

sikai00 committed:
Parser.java: Refactor prepareAdd command

sikai00 committed:
Refactor large test into individual tests

sikai00 committed:
CommandResult.java: Remove old unused constructor

sikai00 committed:
Fix bug with ExitCommand still using old constructor

sikai00 committed:
Make task items have a default, clean format

sikai00 committed:
Make window resizable

sikai00 committed:
Fix alignment in TaskItem

sikai00 committed:
Set minimum resizable width for stage

sikai00 committed:
Merge branch 'A-BetterGui'

sikai00 committed:
Complete A-Personality. A unique personality

sikai00 committed:
Remove unused imports

sikai00 committed:
Complete A-MoreErrorHandling. More error handling

sikai00 committed:
Add corrupted files handling

sikai00 committed:
Storage.java: Remove unnecessary return statement

sikai00 committed:
Tidy up HelpCommand

sikai00 committed:
Add author to header

sikai00 committed:
Complete usage guide, add product screenshot

sikai00 committed:
Complete A-UserGuide. Add a User Guide

sikai00 committed:
Fix command summary table breaking

sikai00 committed:
Resize screenshots

sikai00 committed:
Add link to release

sikai00 committed:
Replace screenshots with better ones

sikai00 committed:
Replace screenshots with even better ones

sikai00 committed:
Fix bug where excess modifications to storage breaks safeguard

sikai00 committed:
Fix storage safeguard on Windows

sikai00 committed:
Fix FXML version error message

sikai00 committed:
Update download link

sikai00 committed:
Add warning to prevent use of delimiter character

sikai00 committed:
Update header comment to fit style convention

sikai00 committed:
Fix typo on event examples

sikai00 committed:
Update README.md
